src/Common/src/System/Net/Security/CertificateHelper.cs

```diff
@@ -33,39 +34,56 @@ internal static X509Certificate2 GetEligibleClientCertificate(X509Certificate2Co
                 return null;
             }
 
-            // Build a new collection with certs that have a private key. We need to do this manually because there is
-            // no X509FindType to match this criteria.
-            // Find(...) returns a collection of clones instead of a filtered collection, so do this before calling
-            // Find(...) to minimize the number of unnecessary allocations and finalizations.
-            var eligibleCerts = new X509Certificate2Collection();
             foreach (X509Certificate2 cert in candidateCerts)
             {
-                if (cert.HasPrivateKey)
+                if (!cert.HasPrivateKey)
                 {
-                    eligibleCerts.Add(cert);
+                    continue;
+                }
+                
+                if (IsValidClientCertificate(cert))
```

Comment:
In terms of validating the client certificate, that is the remote endpoint's job to do that.  That is what the current behavior is today with how which client certificate is ultimately selected to be sent to server. 

This method GetEligibleClientCertificate() hasn't really changed much with this PR other than to check the same stuff as was done with the X509Certificate2Collection.Find() API.  This PR makes those checks more efficient by doing a one-pass thru the list and not allocating new copies of certs.  But in general, this method is simplying try to increase the odds of the server "approving" the client certificate. Any security checks on the client certificate (such as trust, etc.) is validated by the server.

Reply:
And to clarify here, this is NOT the code that SslStream.AsServer uses to validate a received client certificate.  That code lives elsewhere and does full security checks of the received client certificate.

Reply:
I understand that this isn't part of the security boundary. That said, expired or revoked certificates or bad signatures seem pretty unlikely to be accepted by the server (and expiration is pretty common). If you think that's too breaking, you don't need to do anything though.

Reply:
I'd rather not change the behavior like that in this PR.  .NET Framework uses this same algorithm that we currently use here in .NET Core to help prune down the list of eligible client certificates to send to a server.

But it might be something we could consider changing in the future.